smk/gene_abundance.smk

```diff
@@ -662,6 +716,76 @@ rule merge_msamtools_profiles:
         ) >& {log}
         """
 
+rule add_annotation_to_genome_profiles:
+    input:
+        profile  = "{wd}/{omics}/9-mapping-profiles/{minto_mode}/genome_abundances.p{identity}.relabund.prop.tsv",
+        locusmap = "{wd}/DB/{minto_mode}/1-prokka/locus_id_list.txt",
+        taxonomy = "{wd}/DB/{minto_mode}/3-taxonomy/taxonomy.{taxonomy}.{db_version}.tsv"
+    output:
+        mag_profile = "{wd}/output/9-mapping-profiles/{minto_mode}/{omics}.{taxonomy}.{db_version}.p{identity}.tsv",
+        sp_profile  = "{wd}/output/9-mapping-profiles/{minto_mode}/{omics}.{taxonomy}.{db_version}.p{identity}.species.tsv"
+    log:
+        "{wd}/logs/{omics}/9-mapping-profiles/{minto_mode}/{omics}.{taxonomy}.{db_version}.p{identity}.log"
+    wildcard_constraints:
+        taxonomy='gtdb|phylophlan',
+        identity='[0-9]+'
+    resources:
+        mem=10
+    conda:
+        config["minto_dir"]+"/envs/r_pkgs.yml"
+    shell:
+        """
+        time (
+            R --vanilla --silent --no-echo >> {output} <<___EOF___
+
+library(data.table)
+
+# Read tables
+profile  = fread('{input.profile}',  header = TRUE, sep = "\\t")
+locusmap = fread('{input.locusmap}', header = TRUE, sep = "\\t")
+taxonomy = fread('{input.taxonomy}', header = TRUE, sep = "\\t")
+
+# Link locus_tag and mag_id in the profile
+# Keep Unmapped via left-join
+dt = (
+        merge(profile, locusmap, by.x='ID', by.y='locus_id', all.x=TRUE)
+        [is.na(mag_id), mag_id := 'Unmapped']
+     )
+
+# Add taxonomy using mag_id
+# Keep Unmapped via left-join
+# Annotate Unmapped as Unknown
+dt = (
+        merge(dt, taxonomy, by='mag_id', all.x=TRUE)
+        [, lapply(.SD, function(x) ifelse(is.na(x), 'Unknown', x))]
+        [, ID := NULL]
+     )
+setcolorder(dt, 'mag_id')
+
+# Write mag_profile
+fwrite(dt, file = '{output.mag_profile}', row.names = F, col.names = T, sep = "\\t", quote = F)
+
+# Get species profile
+dt = (
+        dt
+        [, mag_id := NULL]
+        [, lapply(.SD, sum), by = c('kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species')]
+        [, taxa_ID := species]
+     )
+
+# Set taxa_ID as first column
```

Comment:
It is philosophical indeed. Both will be acceptable. Having one entity might be less confusing but it reduces information - so there's always pros and cons. 

Current implementation keeps them separate at `MAG` level - when you look at `output/9-mapping-profiles/MAG/metaG.phylophlan.SGB.Jun23.p97.tsv` (I faked an unannotated MAG in tutorial).
```
mag_id  metaG.CD1_36    metaG.CD1_38    kingdom phylum  class   order   family  genus   species
Unmapped        0.0061815495    0.0079989476    Unknown Unknown Unknown Unknown Unknown Unknown Unknown
vae384.MetaSPAdes.k21-99.CD142_146      0.19100249      0.45849847      Unknown Unknown Unknown Unknown Unknown Unknown Unknown
```

But species level file `output/9-mapping-profiles/MAG/metaG.phylophlan.SGB.Jun23.p97.species.tsv` groups them together because they are all coming from unknown species, because `group_by()` and `sum()` happens at `species` column.
```
taxa_ID metaG.CD1_36    metaG.CD1_38    kingdom phylum  class   order   family  genus   species
Unknown 0.1971840395    0.4664974176    Unknown Unknown Unknown Unknown Unknown Unknown Unknown
```

There is one hidden bug though - because the actual `group_by()` is done with all ranks (`by = c('kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species')`) due to `data.table`'s nature. This works when an annotated MAG always has `species` column annotated. Then the `group_by()` variable is always unique. In GTDB and phylophlan, this is the case. If we ever use a taxonomic annotation system that can annotate species as `NA` but other levels with real values, then it will create two entities with `Unknown` for species.
```
taxa_ID metaG.CD1_36    metaG.CD1_38    kingdom phylum  class   order   family  genus   species
Unknown 0.0061815495    0.0079989476    Unknown Unknown Unknown Unknown Unknown Unknown Unknown
Unknown 0.19100249      0.45849847      Bacteria        Bacteroidota    Bacteroidia     Bacteroidales   Bacteroidaceae  Bacteroides     Unknown
```

It looks innocent but could cause trouble if people assume that `taxa_ID` is unique. I assume `R` scripts will fail when there is duplicate values for unique key assumptions.

For now, I am going to leave it like this. I am adding a note for future editors of this code: 862474a

Reply:
Removed species profile output: c909f03